config/options.yml

```diff
@@ -57,6 +58,9 @@ shared:
     cloudflare_bucket: <%= ENV["CLOUDFLARE_BUCKET"] || default_to(cloudflare: :bucket) %>
     google_auth_client_id: <%= ENV["GOOGLE_AUTH_CLIENT_ID"] || default_to(:google_auth_client_id) %>
     google_auth_client_secret: <%= ENV["GOOGLE_AUTH_CLIENT_SECRET"] || default_to(:google_auth_client_secret) %>
+    microsoft_graph_auth_client_id: <%= ENV["MICROSOFT_GRAPH_AUTH_CLIENT_ID"] || default_to(:microsoft_graph_auth_client_id) %>
+    microsoft_graph_auth_client_secret: <%= ENV["MICROSOFT_GRAPH_AUTH_CLIENT_SECRET"] || default_to(:microsoft_graph_auth_client_secret) %>
+    microsoft_graph_scope: <%= ENV["MICROSOFT_GRAPH_SCOPE"] || default_to(:microsoft_graph_scope) %>
```

Comment:
Let’s do: `microsoft: :graph_scope` and similarly for the others, then when someone does bin/rails credentials:edit they can nest all these subkeys under microsoft, e.g.:
```
…
microsoft:
  auth_client_id: xyz
  auth_client_secret: xyz
  graph_scope: xyz
```

(I wish I would have done that for the google keys… I think I added this nesting feature when I was working on cloudflare, right after the google feature. I need to refactor the google one in a way that won’t break it for anyone who has already configured it.)

Reply:
I'll discuss the "microsoft graph" naming here. It seems to be the name microsoft gives to their API service [1], and `microsoft_graph` is the named used by the omniauth provider implementations [2] [3] with names `microsoft_graph` and `microsoft_graph_auth` respectively [4] [5]

So we'll go with `microsoft_graph` as the full name everywhere. One name everywhere for the same thing. I had mix/match of `microsoft` and `microsoft_graph` during development and it led to confusion + failing tests.

But the login button is "Login with Microsoft" for public consumption since no human knows "Microsoft Graph" :) 

* [1] https://learn.microsoft.com/en-us/graph/overview
* [2] https://rubygems.org/gems/omniauth-microsoft_graph
* [3] https://github.com/microsoftgraph/msgraph-sample-rubyrailsapp/blob/main/graph-sample/lib/microsoft_graph_auth.rb (an alternate implementation bundled inside a microsoft sample repo
* [4] https://github.com/synth/omniauth-microsoft_graph/blob/main/lib/omniauth/strategies/microsoft_graph.rb#L10
* [5] https://github.com/synth/omniauth-microsoft_graph/blob/main/lib/omniauth/strategies/microsoft_graph.rb#L10

Reply:
I'll investigate supporting the nested credentials. I didn't know we had that, so will add support for it.

Reply:
I'll be aiming for:

```yaml
microsoft_graph:
  auth_client_id:
  auth_client_secret:
  scope: ...
```

Reply:
Ok, I added support for google + microsoft nested credentials.

```yaml
    google_auth_client_id: <%= ENV["GOOGLE_AUTH_CLIENT_ID"] || default_to(:google, :auth_client_id) || default_to(:google_auth_client_id) %>
    google_auth_client_secret: <%= ENV["GOOGLE_AUTH_CLIENT_SECRET"] || default_to(:google, :auth_client_secret) || default_to(:google_auth_client_secret) %>
    microsoft_graph_auth_client_id: <%= ENV["MICROSOFT_GRAPH_AUTH_CLIENT_ID"] || default_to(:microsoft_graph, :auth_client_id) %>
    microsoft_graph_auth_client_secret: <%= ENV["MICROSOFT_GRAPH_AUTH_CLIENT_SECRET"] || default_to(:microsoft_graph, :auth_client_secret) %>
```